tests/_utils/test_functions.py

```diff
@@ -0,0 +1,1135 @@
+"""Tests for lilypad._utils.functions module."""
+
+import os
+from abc import ABC
+from collections.abc import AsyncIterable, Callable, Iterable
+from typing import Any
+from unittest.mock import AsyncMock, Mock, patch
+
+import pytest
+from jiter import jiter  # pyright: ignore [reportAttributeAccessIssue]
+from mirascope.core import base as mb
+from mirascope.core.base.call_response import _BaseToolT
+from mirascope.core.base.tool import BaseTool
+from openai.types.chat import (
+    ChatCompletion,
+    ChatCompletionMessage,
+    ChatCompletionMessageToolCall,
+    ChatCompletionToolMessageParam,
+)
+from openai.types.chat.chat_completion import Choice
+from pydantic import BaseModel, computed_field
+from pydantic.json_schema import SkipJsonSchema
+
+from lilypad import Message
+from lilypad._utils.functions import create_mirascope_call, inspect_arguments
+from lilypad.server.models import Provider
+
+
+class MockResponse(
```

Comment:
have we already defined this for messages.py? if so, I think maybe we should make this a fixture and put it in a conftest?